### DISCUSSION TITLE
How to choose the right nonconformity score for insurance data?

### DISCUSSION THREAD

**Discussion by burning-cost** [burning-cost/insurance-conformal#4] [Q&A]
I understand the library supports several nonconformity score options. For a pure premium Tweedie model, I'm not clear on when to use each one. The docs mention absolute residuals, normalised residuals, and CQR-based scores.

In my data, severity is very heteroskedastic — large vehicles and high-value postcode areas have much higher variance. Would normalised residuals handle this automatically, or do I need to do something explicit to account for the heteroskedasticity in the conformal step?

**Comment by burning-cost** [comment]
Good question — the choice of nonconformity score is probably the most consequential decision in the conformal setup, and the docs could be more explicit about the tradeoffs.

Here is how to think about it for insurance pure premium models:

**Absolute residuals** `|y - ŷ|`
Use this only if your base model already produces well-calibrated predictions and the residuals are roughly homoskedastic. For insurance data this is almost never true. The interval width will be constant across all risks, which means it will be too wide for low-risk policies and dangerously narrow for high-risk ones.

**Normalised residuals** `|y - ŷ| / σ̂`
This is the right default for heteroskedastic data. The normalisation factor `σ̂` is estimated by a secondary model (by default a CatBoost regressor fit on the absolute residuals from the primary model). This does handle heteroskedasticity automatically — the key is that `σ̂` should be a function of the same features that drive variance, not just a scalar:

```python
from insurance_conformal import TweedieConformalPredictor
from insurance_conformal.scores import NormalisedResidualScore

predictor = TweedieConformalPredictor(
    base_model=catboost_pure_premium_model,
    nonconformity_score=NormalisedResidualScore(
        variance_model="catboost",   # fits a secondary model for σ̂
        variance_features=features   # same features as primary model
    ),
    alpha=0.10
)
predictor.calibrate(cal_df)
```

For your case — high-value postcodes and large vehicles — the `variance_features` should definitely include `vehicle_value`, `sum_insured`, and the postcode band. The secondary model will learn that `σ̂` is higher for these segments.

**CQR score** `max(q_lo - y, y - q_hi)`
This is best when you want tight, adaptive intervals and you are willing to fit two extra quantile models. It tends to outperform normalised residuals when the conditional distribution of losses is highly asymmetric — which is common in severity models but less so in pure premium where zeros are already averaged out.

**My recommendation for your setup:** start with normalised residuals and check the conditional coverage across vehicle value deciles on a test set. If you see systematic undercoverage for high-value vehicles (which is likely), switch to CQR. The CQR intervals will adapt more naturally to the asymmetric loss distribution in the high-value tail.

One practical note: the secondary variance model needs its own train/calibration split. Do not fit `σ̂` on the same data you use for conformal calibration — that introduces a subtle overfitting in the normalisation that inflates coverage in the training region and degrades it elsewhere.
